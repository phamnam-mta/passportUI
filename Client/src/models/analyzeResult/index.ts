import { V3_0_3_AnalyzeResult } from "./v3_0_3";

export type AnalyzeResponse = {
    createdDateTime: string;
    lastUpdatedDateTime: string;
    status: string;
    analyzeResult: V3_0_3_AnalyzeResult;
};

export type StudioAnalyzeResult = {
    apiVersion: string;
    modelId: string;
    stringIndexType: StringIndexType;
    content: string;
    pages: StudioDocumentPage[];
    blocks?: StudioBlock[];
    paragraphs?: StudioParagraph[];
    tables?: StudioDocumentTable[];
    keyValuePairs?: StudioDocumentKeyValuePair[];
    entities?: StudioDocumentEntity[];
    styles?: StudioDocumentStyle[];
    documents?: StudioDocumentResult[];
};

type StringIndexType = "textElements" | "unicodeCodePoint" | "utf16CodeUnit";

export type StudioDocumentPage = {
    pageNumber: number;
    angle: number;
    width: number;
    height: number;
    unit: "pixel" | "inch";
    words: StudioWord[];
    images?: StudioImage[];
    selectionMarks?: StudioSelectionMark[];
    lines?: StudioLine[];
    spans?: StudioSpan[];
    kind?: string;
};

export type StudioDocumentTable = {
    rowCount: number;
    columnCount: number;
    cells: StudioDocumentTableCell[];
    boundingRegions: BoundingRegion[];
    spans: StudioSpan[];
};

export type StudioDocumentKeyValuePair = {
    key: StudioKeyValuePairElement;
    value?: StudioKeyValuePairElement;
    confidence: Confidence;
};

export type StudioDocumentEntity = {
    category: string;
    subCategory?: string;
    content: string;
    boundingRegions: BoundingRegion[];
    confidence: Confidence;
    spans: StudioSpan[];
};

export type StudioDocumentStyle = {
    confidence?: Confidence;
    spans: StudioSpan[];
    isHandwritten?: boolean;
};

export type StudioDocumentResult = {
    docType: string;
    boundingRegions: BoundingRegion[];
    fields: { [fieldName: string]: StudioFieldValue };
    confidence: Confidence;
    spans: StudioSpan[];
};

export type BoundingRegion = {
    pageNumber: number;
    polygon: Polygon;
};

export type StudioWord = {
    content: string;
    polygon: Polygon;
    confidence?: Confidence;
    span?: StudioSpan;
};

export type StudioImage = {
    pageNumber: number;
    confidence?: Confidence;
    span: StudioSpan;
};

export type StudioLine = {
    content: string;
    polygon: Polygon;
    spans: StudioSpan[];
};

export type StudioBlock = {
    kind: string;
    BoundingRegion: BoundingRegion[];
    content: string;
    spans: StudioSpan[];
    confidence: Confidence;
};

export type StudioParagraph = {
    spans: StudioSpan[];
    boundingRegions?: BoundingRegion[];
    content: string;
    role?: string;
};

export type StudioSelectionMark = {
    state: SelectionMarkState;
    polygon: Polygon;
    confidence?: Confidence;
    span?: StudioSpan;
};

export type StudioDocumentTableCell = {
    kind?: DocumentTableCellKind;
    rowIndex: number;
    columnIndex: number;
    rowSpan?: number;
    columnSpan?: number;
    content: any;
    boundingRegions: BoundingRegion[];
    spans?: StudioSpan[];
};

type StudioKeyValuePairElement = {
    content: string;
    boundingRegions: BoundingRegion[];
    spans: StudioSpan[];
};

export type StudioFieldValue = {
    type: FieldValueType;
    valueString?: string;
    valueDate?: string;
    valueTime?: string;
    valuePhoneNumber?: string;
    valueNumber?: number;
    valueInteger?: number;
    valueSelectionMark?: SelectionMarkState;
    valueSignature?: DocumentSignatureType;
    valueCountryRegion?: string;
    valueCurrency?: DocumentCurrencyType;
    valueArray?: StudioFieldValue[];
    valueObject?: { [name: string]: StudioFieldValue };
    valueAddress?: { [name: string]: string };
    content?: string;
    boundingRegions?: BoundingRegion[];
    confidence?: Confidence;
    spans?: StudioSpan[];
};

export type StudioSpan = {
    offset: number;
    length: number;
};

export type Polygon = [number, number, number, number, number, number, number, number];

type DocumentCurrencyType = {
    currencySymbol?: string;
    currency?: string;
    amount?: number;
};

export type SelectionMarkState = "selected" | "unselected";

type Confidence = number;

type DocumentTableCellKind = "content" | "rowHeader" | "columnHeader" | "stub" | "description";

type FieldValueType =
    | "string"
    | "date"
    | "time"
    | "phoneNumber"
    | "number"
    | "integer"
    | "selectionMark"
    | "countryRegion"
    | "currency"
    | "signature"
    | "array"
    | "object"
    | "address";

type DocumentSignatureType = "signed" | "unsigned";

export type SpanPositions = {
    offset: number;
    length: number;
    iconName: string;
    pageNumber?: number;
    tableIndex?: number;
};

export type AnalyzeContentParseResult = {
    documentName: string;
    content: string;
    displayedContent: string;
    positions: Array<{
        column: number;
        lineNumber: number;
        iconName: string;
        shouldDisplayAsIcon: boolean;
        pageNumber?: number;
        tableIndex?: number;
    }>;
};

export type ParsedContentTextBlock = {
    kind?: string | undefined;
    role?: string | undefined;
    content: string;
    boundingRegions?: BoundingRegion[];
};

export type ParsedContentPagedText = {
    [pageNumber: string]: ParsedContentTextBlock[];
};

export type ParsedContentPagedTables = {
    [pageNumber: string]: StudioDocumentTable[];
};

export type ParsedContentPagedSelectionMarks = {
    [pageNumber: string]: StudioSelectionMark[];
};

export type StudioGeneralKeyValue = StudioDocumentKeyValuePair & {
    fieldColor?: string;
};

export type StudioGeneralEntityValue = {
    category: string;
    fieldColor: string;
    valueArray: StudioDocumentEntity[];
};
