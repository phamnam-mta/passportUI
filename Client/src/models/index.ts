export type SplitPaneSizes = {
    analyzeSplitPaneSize: number[];
    labelSplitPaneSize: number[];
    labelTableSplitPaneSize: number[];
};
