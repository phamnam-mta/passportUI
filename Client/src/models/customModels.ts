import { FeatureCategory } from "view/components/imageMap/contracts";
import { Polygon } from "./analyzeResult";

export type CustomModel = {
    modelId: string;
    description: string;
    createdDateTime: string;
    apiVersion?: string;
    tags?: {
        modelType?: string;
        [key: string]: string | undefined;
    };
};

export type PrimitiveField = {
    fieldKey: string;
    fieldType: FieldType;
    fieldFormat: FieldFormat;
};

export type ObjectField = {
    fieldKey: string;
    fieldType: FieldType;
    fields: Field[];
    fieldFormat: FieldFormat;
    visualizationHint?: VisualizationHint;
};

export type ArrayField = {
    fieldKey: string;
    fieldType: FieldType;
    itemType: string;
    visualizationHint?: VisualizationHint;
};

export type Field = PrimitiveField | ObjectField | ArrayField;

export type FieldsWithOrder = Field & { order: number };

export enum FieldType {
    String = "string",
    Number = "number",
    Date = "date",
    Time = "time",
    Integer = "integer",
    SelectionMark = "selectionMark",
    Signature = "signature",
    Array = "array",
    Object = "object",
}

export enum FieldFormat {
    NotSpecified = "not-specified",
    Currency = "currency",
    Decimal = "decimal",
    DecimalCommaSeparated = "decimal-comma-separated",
    NoWhiteSpaces = "no-white-spaces",
    Alphanumeric = "alphanumeric",
    DMY = "dmy",
    MDY = "mdy",
    YMD = "ymd",
}

export enum VisualizationHint {
    Horizontal = "horizontal",
    Vertical = "vertical",
}

export type Labels = {
    [documentName: string]: Label[];
};

export type Label = {
    label: string;
    value: LabelValue[];
    labelType?: LabelType;
};

export type LabelValue = {
    boundingBoxes: Polygon[];
    page: number;
    text: string;
};

export type LabelValueCandidate = {
    boundingBoxes: Polygon[];
    page: number;
    text: string;
    category: FeatureCategory;
    alreadyAssignedLabelName?: string;
};

export enum LabelType {
    Words = "words",
    Region = "region",
}

export type Definitions = {
    [objectName: string]: ObjectField;
};

export enum TableType {
    dynamic = "dynamic",
    fixed = "fixed",
}

export enum HeaderType {
    row = "row",
    column = "column",
}
