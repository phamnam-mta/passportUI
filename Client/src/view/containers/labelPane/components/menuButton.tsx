import * as React from "react";
import { IconButton, IButtonStyles, DirectionalHint } from "@fluentui/react";
import { FieldType } from "models/customModels";

export interface IMenuButtonProps {
    hasSubType?: boolean;
    subType?: FieldType;
    disabled?: boolean;
    onSwitchSubType: (type) => void;
    onDeleteField: () => void;
    onRenameField: () => void;
}

export class MenuButton extends React.PureComponent<IMenuButtonProps> {
    private subTypeOptions = [
        { key: FieldType.String, text: FieldType.String },
        { key: FieldType.Number, text: FieldType.Number },
        { key: FieldType.Date, text: FieldType.Date },
        { key: FieldType.Time, text: FieldType.Time },
        { key: FieldType.Integer, text: FieldType.Integer },
    ];

    private commandOptions = [
        {
            key: "subType",
            text: "Sub type",
            iconProps: { iconName: "DoubleBookmark" },
        },
        {
            key: "rename",
            text: "Rename",
            iconProps: { iconName: "Rename" },
            onClick: this.props.onRenameField,
        },
        {
            key: "delete",
            text: "Delete",
            iconProps: { iconName: "Delete" },
            title: "Delete",
            onClick: this.props.onDeleteField,
        },
    ];

    private menuButtonStyles: IButtonStyles = {
        menuIcon: {
            // This is for hiding the chevron icon, please note that we can't use display: none because
            // it will be overwritten by ms-Icon class.
            width: 0,
            height: 0,
            margin: 0,
            overflow: "hidden",
        },
        rootDisabled: { backgroundColor: "transparent" },
        icon: { fontSize: 14 },
        rootHasMenu: { width: 24, height: 24 },
    };

    private handleSwitchSubType = (type: FieldType) => {
        const { subType, onSwitchSubType } = this.props;
        if (type === subType) {
            return;
        } else {
            onSwitchSubType(type);
        }
    };

    private getSubTypeMenu = () => {
        return this.subTypeOptions.map((option) => ({
            ...option,
            canCheck: true,
            checked: this.props.subType === option.key,
            onClick: () => this.handleSwitchSubType(option.key),
        }));
    };

    private makeCommandMenu = () => {
        return {
            items: this.commandOptions.map((option) => ({
                ...option,
                iconProps: option.iconProps,
                subMenuProps: option.key === "subType" ? { items: this.getSubTypeMenu() } : undefined,
                disabled: option.key === "subType" ? !this.props.hasSubType : false,
            })),
            directionalHint: DirectionalHint.bottomRightEdge,
        };
    };

    public render() {
        return (
            <IconButton
                title="Open menu"
                disabled={this.props.disabled}
                styles={this.menuButtonStyles}
                iconProps={{ iconName: "MoreVertical" }}
                menuProps={this.makeCommandMenu()}
            />
        );
    }
}

export default MenuButton;
