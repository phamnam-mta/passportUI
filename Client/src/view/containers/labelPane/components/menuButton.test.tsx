import * as React from "react";

import { FieldType } from "models/customModels";
import { MenuButton } from "./menuButton";
import { shallow } from "enzyme";

describe("<MenuButton />", () => {
    let baseProps;
    beforeEach(() => {
        baseProps = {
            disabled: false,
            onSwitchSubType: jest.fn(),
            onDeleteField: jest.fn(),
            onRenameField: jest.fn(),
        };
    });

    describe("Rendering", () => {
        it("should render correctly", () => {
            const wrapper = shallow(<MenuButton {...baseProps} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("should render correctly, when there's sub type", () => {
            const props = { ...baseProps, hasSubType: true, subType: FieldType.String };

            const wrapper = shallow(<MenuButton {...props} />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("Event Handling", () => {
        it("should switch sub type, when a sub type is clicked", () => {
            const props = { ...baseProps, hasSubType: true, subType: FieldType.String };

            const wrapper = shallow(<MenuButton {...props} />);
            const button = wrapper.find("CustomizedIconButton") as any;
            // Click number sub type.
            button.prop("menuProps").items[0].subMenuProps.items[1].onClick();

            expect(baseProps.onSwitchSubType).toBeCalledTimes(1);
            expect(baseProps.onSwitchSubType).toBeCalledWith(FieldType.Number);
        });

        it("should not switch sub type, when the same sub type is clicked", () => {
            const props = { ...baseProps, hasSubType: true, subType: FieldType.String };

            const wrapper = shallow(<MenuButton {...props} />);
            const button = wrapper.find("CustomizedIconButton") as any;
            // Click string sub type.
            button.prop("menuProps").items[0].subMenuProps.items[0].onClick();

            expect(baseProps.onSwitchSubType).not.toBeCalled();
        });

        it("should delete field, when delete is clicked", () => {
            const wrapper = shallow(<MenuButton {...baseProps} />);
            const button = wrapper.find("CustomizedIconButton") as any;
            button.prop("menuProps").items[2].onClick(); // Click delete.

            expect(baseProps.onDeleteField).toBeCalledTimes(1);
        });

        it("should rename field, when rename is clicked", () => {
            const wrapper = shallow(<MenuButton {...baseProps} />);
            const button = wrapper.find("CustomizedIconButton") as any;
            button.prop("menuProps").items[1].onClick(); // Click rename.

            expect(baseProps.onRenameField).toBeCalledTimes(1);
        });
    });
});
