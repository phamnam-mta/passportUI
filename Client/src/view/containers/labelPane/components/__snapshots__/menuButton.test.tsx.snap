// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuButton /> Rendering should render correctly 1`] = `
<CustomizedIconButton
  disabled={false}
  iconProps={
    Object {
      "iconName": "MoreVertical",
    }
  }
  menuProps={
    Object {
      "directionalHint": 6,
      "items": Array [
        Object {
          "disabled": true,
          "iconProps": Object {
            "iconName": "DoubleBookmark",
          },
          "key": "subType",
          "subMenuProps": Object {
            "items": Array [
              Object {
                "canCheck": true,
                "checked": false,
                "key": "string",
                "onClick": [Function],
                "text": "string",
              },
              Object {
                "canCheck": true,
                "checked": false,
                "key": "number",
                "onClick": [Function],
                "text": "number",
              },
              Object {
                "canCheck": true,
                "checked": false,
                "key": "date",
                "onClick": [Function],
                "text": "date",
              },
              Object {
                "canCheck": true,
                "checked": false,
                "key": "time",
                "onClick": [Function],
                "text": "time",
              },
              Object {
                "canCheck": true,
                "checked": false,
                "key": "integer",
                "onClick": [Function],
                "text": "integer",
              },
            ],
          },
          "text": "Sub type",
        },
        Object {
          "disabled": false,
          "iconProps": Object {
            "iconName": "Rename",
          },
          "key": "rename",
          "onClick": [MockFunction],
          "subMenuProps": undefined,
          "text": "Rename",
        },
        Object {
          "disabled": false,
          "iconProps": Object {
            "iconName": "Delete",
          },
          "key": "delete",
          "onClick": [MockFunction],
          "subMenuProps": undefined,
          "text": "Delete",
          "title": "Delete",
        },
      ],
    }
  }
  styles={
    Object {
      "icon": Object {
        "fontSize": 14,
      },
      "menuIcon": Object {
        "height": 0,
        "margin": 0,
        "overflow": "hidden",
        "width": 0,
      },
      "rootDisabled": Object {
        "backgroundColor": "transparent",
      },
      "rootHasMenu": Object {
        "height": 24,
        "width": 24,
      },
    }
  }
  title="Open menu"
/>
`;

exports[`<MenuButton /> Rendering should render correctly, when there's sub type 1`] = `
<CustomizedIconButton
  disabled={false}
  iconProps={
    Object {
      "iconName": "MoreVertical",
    }
  }
  menuProps={
    Object {
      "directionalHint": 6,
      "items": Array [
        Object {
          "disabled": false,
          "iconProps": Object {
            "iconName": "DoubleBookmark",
          },
          "key": "subType",
          "subMenuProps": Object {
            "items": Array [
              Object {
                "canCheck": true,
                "checked": true,
                "key": "string",
                "onClick": [Function],
                "text": "string",
              },
              Object {
                "canCheck": true,
                "checked": false,
                "key": "number",
                "onClick": [Function],
                "text": "number",
              },
              Object {
                "canCheck": true,
                "checked": false,
                "key": "date",
                "onClick": [Function],
                "text": "date",
              },
              Object {
                "canCheck": true,
                "checked": false,
                "key": "time",
                "onClick": [Function],
                "text": "time",
              },
              Object {
                "canCheck": true,
                "checked": false,
                "key": "integer",
                "onClick": [Function],
                "text": "integer",
              },
            ],
          },
          "text": "Sub type",
        },
        Object {
          "disabled": false,
          "iconProps": Object {
            "iconName": "Rename",
          },
          "key": "rename",
          "onClick": [MockFunction],
          "subMenuProps": undefined,
          "text": "Rename",
        },
        Object {
          "disabled": false,
          "iconProps": Object {
            "iconName": "Delete",
          },
          "key": "delete",
          "onClick": [MockFunction],
          "subMenuProps": undefined,
          "text": "Delete",
          "title": "Delete",
        },
      ],
    }
  }
  styles={
    Object {
      "icon": Object {
        "fontSize": 14,
      },
      "menuIcon": Object {
        "height": 0,
        "margin": 0,
        "overflow": "hidden",
        "width": 0,
      },
      "rootDisabled": Object {
        "backgroundColor": "transparent",
      },
      "rootHasMenu": Object {
        "height": 24,
        "width": 24,
      },
    }
  }
  title="Open menu"
/>
`;
